src/proto/ping.rs

```diff
@@ -0,0 +1,102 @@
+//! MC Server List Ping protocol.
+//!
+//! Reference: http://wiki.vg/Server_List_Ping
+
+use std::fs::File;
+use std::io::prelude::*;
+use std::io;
+use std::net::TcpStream;
+use std::ops::Sub; // Sub for Timespec
+use std::path::Path;
+
+use packet::{self, PacketRead, PacketWrite};
+
+use rustc_serialize::base64::{ToBase64, STANDARD};
+use rustc_serialize::json;
+use time;
+
+#[derive(RustcDecodable, RustcEncodable)]
+pub struct Version {
+    pub name: String,
+    pub protocol: i32,
+}
+
+#[derive(RustcDecodable, RustcEncodable)]
+pub struct Players {
+    pub max: i32,
+    pub online: i32,
+    // sample is optional
+}
+
+#[derive(RustcDecodable, RustcEncodable)]
+pub struct Description {
+    pub text: String,
+}
+
+/// Response sent to clients as JSON.
+#[derive(RustcDecodable, RustcEncodable)]
+pub struct Response {
+    pub version: Version,
+    pub players: Players,
+    pub description: Description,
+    pub favicon: String,
+}
+
+// FIXME(toqueteos): This is yelling to be a method of a Server struct or
+// something more useful. We need the Handshake's `next_state` field in order
+// to perform login for a player.
+pub fn handler(mut stream: &mut TcpStream) -> io::Result<()> {
+    // C->S: Status Request packet
+    let sr = match try!(packet::status::serverbound::Packet::read(stream)) {
+        packet::status::serverbound::Packet::StatusRequest(sr) => sr,
+        wrong_packet => panic!("Expecting status::serverbound::StatusRequest packet, got {:?}", wrong_packet)
+    };
+
+    // S->C: Status Response packet
+    let mut file = try!(File::open(&Path::new("favicon.png")));
+    let mut contents = Vec::new();
+    try!(file.read_to_end(&mut contents));
+    let favicon = contents.as_slice().to_base64(STANDARD);
+    // FIXME(toqueteos): Micro-optimization? We could totally drop JSON
+    // encoding and just replace player values (online & max) with format! all
+    // other values are static.
+    let resp = Response{
+        version: Version{
+            name: "1.8.1".to_string(),
+            protocol: 47,
+        },
+        players: Players{
+            // FIXME(toqueteos): This is value should be a internal counter of server
+            online: 0,
+            // FIXME(toqueteos): This is value read from server.properties file
+            max: 20,
+        },
+        description: Description{
+            text: "With custom favicons! Woot :D".to_string(),
+        },
+        favicon: format!("data:image/png;base64,{:}", favicon),
+    };
+    let sr = packet::status::clientbound::StatusResponse{
+        response: json::encode(&resp).unwrap()
```

Comment:
No, just impl `Protocol` for the `Response` struct and use that as the `StatusResponse` packet's field.